platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/CrossProjectModelAccessPattern.kt

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache
+
+import org.gradle.api.internal.project.ProjectInternal
+
+
+internal
+enum class CrossProjectModelAccessPattern {
+    DIRECT,
+    CHILD,
+    SUBPROJECT,
+    ALLPROJECTS
+}
+
+
+internal
+data class CrossProjectModelAccessInstance(
+    val pattern: CrossProjectModelAccessPattern,
+    val relativeTo: ProjectInternal,
```

Comment:
💭  Not 100% related, but since this PR is overhauling:
I see `relativeTo` name is using in `ProblemReportingCrossProjectModelAccess` as an opposition to `referrer`. Maybe it would be more expressive, if we have a symmetry in `referrer <-> referent`?

Reply:
I would avoid this because it makes the names too similar and easy to confuse. Even if the tests would catch it, the code would be extra mental load when reading the code.

Personally, even using `referrer` for me is a head-scratcher. I have to stop every time to reassess that it is actually the right thing. Maybe I just don't like the word 😆